### ISSUE TITLE
Offline embeddings without PyTorch — a FastEmbed/ONNX provider (856 MB → 184 MB, identical vectors)

### ISSUE THREAD

**Issue by LuigiNicaPRO** [jgravelle/jdocmunch-mcp#126]
`sentence-transformers` is currently the only offline embedding provider, and it drags PyTorch in
with it. On one real install that is **730 MB of an 856 MB venv** — torch 506, scipy 82,
transformers 59, sklearn 34, numpy 26, tokenizers 9, sentence_transformers 6, huggingface_hub 6,
safetensors 2.

That weight is the root cause behind three issues already closed here:

- **#110** — provider initialized eagerly at MCP startup, ~7.6 s added to every start.
- **#118** — a numpy C-extension import wedges indefinitely inside the running server.
- **#119** — the child-process workaround #118 forced.

FastEmbed runs the **same model** on ONNX Runtime with no torch, no C-extension load of that class,
and a startup measured in well under a second.

### The measurement that makes this a drop-in rather than a migration

`sentence-transformers/all-MiniLM-L6-v2` is in FastEmbed's catalogue at the same 384 dims. I embedded
8 real strings with both backends:

```
cosine, matched pairs      : 1.00000000  (all 8)
worst element-wise |Δ|     : 1.63e-7     (float32 rounding)
CONTROL, mismatched pairs  : 0.008 – 0.148
```

The control is the part that matters — a comparison that always returns 1.0 would look identical to
this one. It discriminates, so the identity is real.

**Consequence: an existing corpus does not need re-embedding.** I ran this against a live 31,339-section
index; the two probe queries I baselined beforehand came back with the same ids in the same order,
largest score delta `1.5e-7`.

### One sharp edge, if you take this

`cache.py:109` validates the sidecar header on **provider AND model**, and `provider.py:680/751` pass
the raw `get_provider_name()` return. So a provider *named* `fastembed` writes a header that
mismatches the stored `sentence-transformers`, `cache.load()` returns `{}`, and every vector is
silently discarded and recomputed — a very expensive no-op that looks like success.

The identity header describes the **vectors**, not the runtime that produced them. Since the vectors
are identical, the name written into the header should stay stable across a runtime swap. I handled
it by normalizing that one local *after* `_get_provider()` has resolved, so the header stays put while
FastEmbed still does the embedding. A cleaner upstream fix might separate "provider" from
"vector family" in the header outright.

Also worth knowing: `get_provider_name()` is a closed if-chain, so registering in `_PROVIDER_FACTORIES`
alone is not enough — an unrecognised value falls through to auto-detect and never reaches the dict.

### Offer

I have this running locally as a patch against v1.136.1 — four anchored edits to `provider.py`
(selection branch, provider class + registration, cache-key signature, header normalization). Happy
to open a PR if you want it, or to hand over the provider class and let you wire it your own way.
The repo carries `NOASSERTION` for a license, so tell me what terms you want a contribution under.

FastEmbed resolves cleanly on cp314/arm64: `fastembed 0.8.0` + `onnxruntime 1.29.0`, 11 packages,
no torch.

**Comment by jgravelle** [comment]
Accepted — please open the PR. Terms first, then four things I want in scope.

**License.** `NOASSERTION` is GitHub failing to classify `LicenseRef-jDocMunch-Dual-Use`, not an absent license; `LICENSE` is in the tree. Contributions come in under that dual-use license via the CLA at https://cla-assistant.io/jgravelle/jdocmunch-mcp — cla-assistant prompts automatically and it takes about 30 seconds. Per CONTRIBUTING.md the signing window is 24 hours from the PR being reviewed and green. If it expires we implement it ourselves and credit you in the CHANGELOG, the release notes and the close comment, so the window decides whose commit it is and nothing else. If 24 hours does not work, say so on the PR and we will hold it.

Every claim above checks out against v1.136.1. `get_provider_name()` (`provider.py:269`) is a closed if-chain and `_PROVIDER_FACTORIES` (`:517`) is only ever reached with a name that chain already returned. `cache.load` compares `provider` and `model` for exact equality (`cache.py:109-118`). And there is no dim backstop underneath it: `_provider_identity` returns `(model, None)` for sentence-transformers, and `dim=None` is a wildcard in that same comparison.

That last point makes the header question sharper than your writeup has it.

### 1. Normalize per MODEL, not per provider

Blanket normalization of that local is the part I would not merge. `JDOCMUNCH_ST_MODEL` is user-settable, and with dim wildcarded nothing downstream can catch a disagreement. Point it at a model FastEmbed either does not carry or carries with different weights or pooling, and a normalized header writes `sentence-transformers` over vectors that are not sentence-transformers' vectors for that model. `cache.load()` then matches, the two derivations merge into one sidecar, and search ranks across both. That is jdoc#111's shape, and it is worse than the recompute you are avoiding: a full re-embed is expensive and observable, this is cheap and invisible.

So the alias should be an explicit allow-list of model ids measured identical across the two runtimes, starting with `sentence-transformers/all-MiniLM-L6-v2`. Inside it, write `sentence-transformers` and reuse the sidecar. Outside it, write `fastembed` and re-embed. Fail closed — the expensive correct path is the right default for a model nobody has measured.

Your mismatched-pair control is the half of that measurement I want in `tests/`, not the 1.00000000. Which raises a question I would rather hear your read on than decide alone: a test that skips everywhere is a test that never ran (we carried six of those here for months without noticing), so putting `fastembed` in the dev group is what makes the assertion execute — and also puts a model download in every CI job.

### 2. `_hf_model_cached()` is shaped to the ST cache layout

`models--<org>--<name>/snapshots` under the HF hub root, `provider.py:188-205`. It exists because of #110: an uncached model means warmup is skipped, which is what removed the 30-second-connect-timeout outage. FastEmbed keeps its own cache. Under a FastEmbed provider that probe is answering about a directory the runtime is not using, in whichever direction it happens to answer, so it needs a provider-aware branch or #110's logic misfires quietly.

### 3. Do not inherit the #118 machinery, and do not assume it is unnecessary

`_get_provider()` gates on `_sentence_transformers_imports_cleanly()`, a probe subprocess that must not run for a provider that never imports the package. Separately: #118's wedge was `LdrLoadDll` inside OpenBLAS's `DllMain`, reached through the numpy/scipy/torch chain. onnxruntime is a different DLL set, so torch's absence is an argument and not evidence. I would keep the worker path reachable for FastEmbed rather than declaring it moot, unless you have a Windows measurement that says otherwise.

### 4. Auto-detect precedence

If both are importable my inclination is FastEmbed wins, with an explicit `JDOCMUNCH_EMBEDDING_PROVIDER=sentence-transformers` still overriding. Vector-identical on the allow-listed model makes that a tightening rather than a wire change. Disagree if you see it differently.

### Disclosure

FastEmbed fetches model weights on first use. This package was PyPI-quarantined once over an undisclosed background behavior, so README's "Background behavior, fully disclosed" section is a compliance surface here — a new extra that reaches the network is written up there before it ships, not after.

PR scope: your four anchored edits, plus the model allow-list, the cache-probe branch, and the identity test carrying your control. `fastembed` as an optional extra, never a hard dependency.

Thanks for the control, and for reading `cache.py:109` before filing rather than after. Both #14 and #15 were the same standard.

**Comment by jgravelle** [comment]
Shipped in [v1.137.0](https://github.com/jgravelle/jdocmunch-mcp/releases/tag/v1.137.0), live on PyPI and the MCP registry. `pip install jdocmunch-mcp[fastembed]`.

Your 24-hour window for the PR expired, so we implemented it — the finding, the four anchored edits and the design are credited to you in the CHANGELOG, the release notes and here. If you would still rather own the commit, say so and we will take a follow-up PR from you for anything in this area.

What went in beyond your patch:

**The alias keys on the model, not the provider.** Your normalization is unconditional, and there is no dim backstop under it — `_provider_identity` returns `dim=None` for both offline providers and the cache reads that as a wildcard. Since `JDOCMUNCH_ST_MODEL` is user-settable, a blanket rename writes `sentence-transformers` over vectors some other model produced, `cache.load` then matches, the two derivations merge into one sidecar, and search ranks across both. So `_FASTEMBED_ST_EQUIVALENT_MODELS` is an explicit per-model allow-list holding `sentence-transformers/all-MiniLM-L6-v2` today, failing closed on an unlisted model, on a divergent `JDOCMUNCH_ST_MODEL`, and on an empty model.

Two details you would hit implementing it: the alias writes the sentence-transformers **spelling** (`all-MiniLM-L6-v2`), not the canonical hub id, because the header is an exact string match against a file that already exists — and the dim stays `None`, because every existing sentence-transformers sidecar stores `None` and an active 384 compares unequal and purges the file the alias exists to reuse.

**`check_embedding_drift` is how a model gets onto that list.** Capture a canary under one runtime, re-run it under the other, read `max_drift`. If you want to propose another model, that measurement is the whole submission.

**The cache probe.** `_st_model_is_cached()` reads the HF hub layout, and #110's warmup gate uses it to decide whether a load would stall the MCP handshake on a download. FastEmbed downloads elsewhere, so the shared probe reports "cached" for a machine whose HF cache holds the model for torch while FastEmbed still has to fetch it. A populated HF cache is now explicitly not evidence.

**The sentence-transformers probe and the embed worker stay off for FastEmbed.** Both are about torch. onnxruntime loads a different DLL set, so #118 is an argument here rather than evidence — if FastEmbed wedges the same way on Windows, that is a new issue and the worker is the fix.

Auto-detect prefers FastEmbed when both offline providers are installed; `JDOCMUNCH_EMBEDDING_PROVIDER=sentence-transformers` is the way back.

On the open question we put to you: `fastembed` is **not** in the dev group. It would make the identity tests execute, but it also adds a model download to every CI job, and the tests that matter here assert on the header triple rather than on vectors. If you have a view, the issue tracker is open.

42 tests, 41 failing against the pre-fix tree. Suite 2646 passed / 6 skipped, all 11 CI jobs plus Replay green at `d27a1dc`.
